Mathlib/GameTheory/Auction/Basic.lean

```diff
@@ -0,0 +1,192 @@
+
+## Main Definitions
+
+- `maxBid`: Function that computes the highest bid given a bidding function.
+- `winner`: Identifies the winner of the auction as the bidder with the highest bid.
+- `maxBidExcluding`: Function that computes the highest bid excluding a given participant.
+- `Dominant`: Defines a dominant strategy in the context of a auction.
+- `secondPrice`: Computes the second highest bid in the auction.
+- `Secondprice.utility`: Computes the utility of each bidder based on the outcome of the auction.
+- `Firstprice.utility`: Computes the utility for a first price auction.
+
+## Main Results
+
+- `utility_nneg`: Utility is non-negative if the bid equals the valuation in second price auction.
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy in second price auction.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in first price auction.
+
+## Helping Lemmas
+
+- `eq_winner_of_bid_gt`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_maxBid`: There exists a participant whose bid matches the highest bid
+- `bid_winner_eq_maxBid`: The winner's bid is the highest.
+- `Secondprice.utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `Secondprice.utility_loser`: If `i` does not win, their utility is 0.
+- `bid_le_bid_winner`: The winner's bid is greater than or equal to all other bids.
+- `maxBidExcluding_eq_maxBid_if_loser`: If `i` does not win, the highest bid excluding `i` equals
+   the highest bid.
+- `Firstprice.utility_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `Firstprice.utility_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxBid` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+/- `v` is the valuation for participants `I` -/
+variable {I : Type*} [Fintype I] [Nontrivial I] (v : I → ℝ)
+
+namespace Auction
+
+/- `b` is the bidding function that maps participant to a real number bid. -/
+variable (b : I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxBid : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
+
+/-- There exists a participant `i` whose bid equals the highest bid. -/
+lemma exists_maxBid : ∃ i : I, b i = maxBid b := by
+   obtain ⟨i, _, h2⟩ := Finset.exists_mem_eq_sup' Finset.univ_nonempty b
+   exact ⟨i, symm h2⟩
+
+/-- winner: The participant with the highest bid. -/
+noncomputable def winner : I := Classical.choose (exists_maxBid b)
```

Comment:
Relating to @YaelDillies' comment: I think the crux of the difficulty of this PR is that it introduces new definitions like `maxBid` and `winner` which are really just unusual names for functions which could be called `max` and `someArgMax`.

Yury's suggestion to abstract further provides another solution to this issue.

Reply:
These names are absolutely not unusual in the field.

Reply:
Sure but mathlib is not meant to be super-tailored to a specific field. In particular, we definitely don't want to have a differently-named-but-otherwise-identical function for each field that uses it! One solution here is to use notation instead. 

Also note that this is basically the only problem with this PR, so I think it is worth solving this way.

Reply:
Two points:
1. Notation that looks like a def name could result people applying tactics with that name and failing.

2. I see a lot of value in developing this theory and a lot of TCS, econ, physics etc (and this is definitely TCS). We are not accustomed to the whole "generalise as much as possible" philosophy of math and thus don't have standard battle tested definitions. In a way this works bette for us as these abstractions only add a layer of unnecessary and uninsightful jargon. Thus rushing a definition that may or may not be ergonomic into mathlib might result in a suboptimal outcome. 

3. As Yury points out, there are likely to be no downstream dependencies of these files outside the folder, and even inside them, some experimentation is going to be needed, and different areas of game theory are going to use different variations of the same conceptual definition.

Reply:
To give one example, I am quite certain that utility functions are going to have a multitude of definitions.

Reply:
A similar opinion has been expressed by Kim for another TCS formalisation: https://leanprover.zulipchat.com/#narrow/stream/252551-graph-theory/topic/Labeled.20Transition.20Systems/near/472581943